docs/guides/dropout.rst

```diff
@@ -0,0 +1,147 @@
+
+  seed = 0
+  root_key = jax.random.PRNGKey(seed=seed)
+  main_key, params_key, dropout_key = jax.random.split(key=root_key, num=3)
+
+Each of the Flax PRNG streams has a name, such as the ``'params'`` stream for
+initializing parameters, and ``'dropout'`` for dropout.
+
+Defining a model with ``Dropout``
+*********************************
+
+To create a model, subclass :meth:`flax.linen.Module`, and then use
+:meth:`flax.linen.Dropout` to add a dropout layer. Recall that
+:meth:`flax.linen.Module` is the
+`base class for all neural network modules <https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#module>`__
+, and all layers and models are subclassed from it.
+
+In :meth:`flax.linen.Dropout`, the ``deterministic`` argument is ``None`` by
+default.
+
+* If it's set to ``False``, the inputs are scaled by ``1 / (1 - dropout_rate)`` and masked.
+* When it's ``True``, no mask is applied (the dropout is turned off).
+
+Flax provides an implicit way of handling PRNG key streams via Flax
+:meth:`flax.linen.Module`'s :meth:`flax.linen.Module.make_rng` helper function.
+It allows the code in Flax ``Module``s (or their sub-``Module``s) to
+"pull PRNG keys". ``make_rng`` guarantees to provide a unique key each time you
+call it.
+
+**Note:** Internally, :meth:`flax.linen.Dropout` makes use of :meth:`flax.linen.Module.make_rng`
+to create a key for dropout. You can check out the
+`source code <https://github.com/google/flax/blob/5714e57a0dc8146eb58a7a06ed768ed3a17672f9/flax/linen/stochastic.py#L72>`__.
+
+* Every time :meth:`Module.make_rng` is called (which is done implicitly in
+  :meth:`flax.linen.Dropout`), you get a new PRNG key split from the main/root pseudorandom
+  number generator (PRNG) key.
+* :meth:`Module.make_rng` guarantees full reproducibility.
+
+.. codediff::
+  :title_left: No Dropout
+  :title_right: With Dropout
+  :sync:
+
+  class MyModel(nn.Module):
+    num_neurons: int
+    training: bool
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Dense(self.num_neurons)(x)
+
+
+
+      return x
+
+  ---
+  class MyModel(nn.Module):
+    num_neurons: int
+    training: bool
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Dense(self.num_neurons)(x)
+      # Set the dropout layer with a rate of 50%. #!
+      # When the `deterministic` flag is `True`, dropout is turned off. #!
+      x = nn.Dropout(rate=0.5, deterministic=not self.training)(x) #!
+      return x
+
+After adding ``Dropout`` to your model, instantiate the model with
+``training=True``.
+
+For example, when creating a ``TrainState``, the code will include::
+
+  my_model = MyModel(num_neurons=3, training=True)
+
+Initializing the model
+**********************
+
+The PRNG key provided to :meth:`flax.linen.init()` is the one that seeds the
+``'params'`` PRNG key stream.
+
+Initialize the model with the ``params'`` key::
+
+  x = jax.random.uniform(key=main_key, shape=(3, 4, 4))
+  variables = my_model.init(params_key, x)
```

Comment:
This is not correct. If you construct `my_model` as you did in the previous section, then you should pass an RNG for `dropout` here. I would also explain that you don't need to pass it if Dropout is disabled.

Maybe you can use a side-by-side diff view here as well?

```suggestion
When applying Dropout, you should pass the RNG key for it to :meth:`flax.linen.init()` with name `dropout`. If do you not apply Dropout, you don't have to pass it. Since you are then only passing the RNG for initializing parameters, you can abbreviate ``{'params': params_rng}`` with simply ``params_rng`` (see :meth:`flax.linen.init()` for more details).

.. codediff::
  :title_left: Dropout disabled
  :title_right: Dropout enabled
  :sync:
  
  rng = random.PRNGKey(0)
  data_rng, params_rng, rng = random.split(rng, 3)
  x = jax.random.uniform(data_rng, shape=(3, 4, 4))
  
  my_model = MyModel(num_neurons=3, training=False)
  variables = model_nodropout.init(params_rng, x)

  ---
  rng = random.PRNGKey(0)
  data_rng, params_rng, dropout_rng, rng = random.split(rng, 4)
  x = jax.random.uniform(data_rng, shape=(3, 4, 4))
  init_rngs = {'params': params_key, 'dropout': dropout_key}
  my_model = MyModel(num_neurons=3, training=True)
  variables = model_dropout.init(init_rngs, x)

```

Reply:
Sounds good!

With some minor changes:

```
- If you enable :meth:`flax.linen.Dropout` in your model, you need to pass the
  ``'dropout'`` PRNG key together with the ``'params'`` key.
- If you have ``Dropout`` disabled, the ``'params'`` PRNG key is enough for
  :meth:`flax.linen.init()`.

Refer to the :meth:`flax.linen.init()` API docs for more details.

.. codediff::
  :title_left: Dropout disabled
  :title_right: Dropout enabled
  :sync:

  my_model_without_dropout = MyModel(num_neurons=3, training=False)

  x = jax.random.uniform(key=main_key, shape=(3, 4, 4))
  variables = my_model.init(params_rngs, x)
  ---
  my_model_with_dropout = MyModel(num_neurons=3, training=True) #!
  init_rngs = {'params': params_key, 'dropout': dropout_key} #!
  x = jax.random.uniform(key=main_key, shape=(3, 4, 4))
  variables = my_model.init(init_rngs, x) #!
```

Reply:
Revamped `init()`, etc, no longer using `  init_rngs = {'params': params_key, 'dropout': dropout_key}`.